### DISCUSSION TITLE
Is there a way to immediately handle error when `retry` is enabled?

### DISCUSSION THREAD

**Discussion by aak-lear** [TanStack/query#4581] [Q&A] [answered]
How can I immediately handle error on first failed request? When `retry` option is enabled, `useQuery` ignores all errors while retrying to fetch data.

Only when retry attempts is finished, `useQuery` changes all error states like `isError`, `isLoadingError`, `isRefetchError` and calls `onError` callback. Is there a way to immediately handle error on the first try?

#### Use case

When session is expired, and client send request to backend, there will be an error 401 in the response.
In that case I know that all retry attempts on this request will get the same error and I want to handle this error immediately. For example, to refresh session or redirect to "Session Expired" page. Otherwise user will be forced to wait while retry cycle ends.

**Comment by TkDodo** [comment, marked as answer]
you can set `retry` to a function and return false if the error is a 401 error to _not_ retry.

this is documented here: https://tanstack.com/query/v4/docs/guides/query-retries

> Setting retry = (failureCount, error) => ... allows for custom logic based on why the request failed.

**Reply by aak-lear** [reply]
Thank you!
